(SLV-414) Add scale test script

johnduarte opened the pull request:
This commit adds a bash script for automating running scale testing
against the defined reference architectures defined in the PE docs

https://puppet.com/docs/pe/latest/hardware_requirements.html

This script will run a single iteration of the "10 or fewer" architecture.  It
will then run the following iterations on three different EC2 instance
definitions to test the scale of the "Up to 4,000" architecture options.

1. Cold start run provisioning the EC2 instances.
1. 4 more cold start iterations on the instance created above.
1. 5 warm start iterations on the instances created above.

The "Monolithic with compilers" architecture is not executed.

johnduarte commented:
This script has not been adequately tested and may cause you to have a bad day.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+}
+
+# Link scale results to nginx content location
+function link_results () {
```
With the latest gplt_runner_setup  linking won't be needed as long as the repo is run from $HOME/gatling

johnduarte replied:
Updated to use `$HOME/gatling` as the primary working directory.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+
+# Prepare checkout of puppet-gatling-load-test
+function prep_gplt () {
+    cd "$HOME/gplt" || exit 1
```
change to $HOME/gatling to take advantage of the latest gplt_runner_setup automatically making everything in there available to nginx

johnduarte replied:
Updated to use `$HOME/gatling` as the primary working directory.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+tune=false
+# Global settings
+export BEAKER_INSTALL_TYPE=pe
+export BEAKER_PE_DIR=http://enterprise.delivery.puppetlabs.net/archives/releases/2019.1.0
```
This should probably use the inputed pe version (if it is a release version).  And if it isn't, the user has to give one... which I don't see a way for them to do since this is hardcoding it.

johnduarte replied:
Logic added to:
1. Interpret whether the given PE_VERSION should be a release or a build
1. Check if tarballs exist for that version

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+mkdir "$HOME/gplt"
+
+echo "========================================================================"
+echo "Testing Ref Arch 1: Monolithic deploy for 10 or fewer nodes"
```
Mono is no more.  It is Standard or Standard Ref Arch.  Also, "for trail use" or something that doesn't quote a number of nodes.

johnduarte replied:
Updated with new terms.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+
+    run_type="cold"
+    ec2=$ABS_AWS_MOM_SIZE
+    test="slv-414-ref1-$ec2"
```
ticket number needs to be configurable, and not locked into being a ticket number... more like a prefix.

johnduarte replied:
Added -i|--run-id option to allow the user to specify an optional prefix.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+    run_type="cold"
+    ec2=$ABS_AWS_MOM_SIZE
+    test="slv-414-ref1-$ec2"
+    cmd="bundle exec rake autoscale_$run_type > \"$test-$run_type.log\""
```
created cmd, but didn't use it.

johnduarte replied:
The `cmd` variable is used by the debug flag.  I couldn't figure out how to get a string and a command out of the same variable. :punting:.

RandellP replied:
isn't it just
eval $cmd
?

johnduarte replied:
Alas, due to the "meta-variable", this did not evaluate the inner variables properly.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+
+
+echo "========================================================================"
+echo "Testing Ref Arch 2: Monolithic deploy for up to 4,000 nodes"
```
Replace Mono with Standard.  Remove hardcoded node assumption since that is what we are testing...

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+    declare -a ec2_types=("c5.xlarge" "c5.2xlarge" "c5.4xlarge")
+    for i in {0..4}; do
+        wait
+        [[ $i = 0 ]] && task="autoscale_$run_type" || task="autoscale_provisioned_$run_type"
```
not sure how this does what it does...  assuming it works, add a comment.

RandellP replied:
still an open question

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+        for ec2 in "${ec2_types[@]}"; do
+            if [[ $ec2 == "c5.xlarge" ]]; then
+                if [[ $tune == "false" ]]; then
+                    export PUPPET_GATLING_SCALE_BASE_INSTANCES=2000
```
These numbers should be extracted into a better data type or a set of default variables so that they are all in one place in just a few lines and able to be seen on one page.

johnduarte replied:
These variables have been grouped at the beginning of the script.  The user can modify them prior to run if desired.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+                export PUPPET_GATLING_SCALE_BASE_INSTANCES=3800
+            fi
+            export ABS_AWS_MOM_SIZE="$ec2"
+            test="slv-414-ref2-$ec2-tune-$tune"
```
prefix hardcode... and since ref2 appears to be meaningful, it should have a meaningful name...  like standard or trial...

johnduarte replied:
Added -i|--run-id option to allow the user to specify an optional prefix.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+            if [ -z $DEBUG ]; then
+                prep_gplt "$test"
+                link_results "$test" "$BEAKER_PE_VER"
+                (bundle exec rake $task > "$test-$run_type-$i.log") &
```
use $cmd

johnduarte replied:
It is used in the debug output

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+            cmd="bundle exec rake $task > \"$test-$run_type-$i.log\""
+            if [ -z $DEBUG ]; then
+                prep_gplt "$test"
+                link_results "$test" "$BEAKER_PE_VER"
```
not needed in new setup

johnduarte replied:
Deleted

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,230 @@
+    cd "$HOME/gplt" || exit 1
+    mkdir "$1"
+    cd "$1" || exit 1
+    git clone git@github.com:puppetlabs/gatling-puppet-load-test
```
instead of creating a dir and putting gatling under it, why not just clone the gatling repo to the dir you are creating directly.

johnduarte replied:
Updated to clone into the parameterized directory name directly without created doubly nested directory structure.

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,330 @@
+# AWSSIZE: [L|XL|2XL|4XL] (only used with BASE_INSTANCES)
+# START: [COLD|WARM] (only used with BASE_INSTANCES)
+# TUNING: [TUNED|UNTUNED] (only used with BASE_INSTANCES)
+# TUNING: [TUNED|UNTUNED] (only used with BASE_INSTANCES)
```
duplicate line

johnduarte replied:
Duplicate line deleted

RandellP commented on util/scale_run.sh:
```diff
@@ -0,0 +1,330 @@
+echo "PE_MINOR=$PE_MINOR"
+echo "PE_PATCH=$PE_PATCH"
+echo "PE_BUILD=$PE_BUILD"
+#if ! [[ $PE_VERSION =~ ^[0-9]+\.[0-9]+\.[0-9]+((-[0-9a-zA-Z-]+)*)?$ ]]; then
```
if intentionally leaving this here, comment on why.

johnduarte replied:
This line has been deleted

johnduarte commented:
@RandellP I have added a doc for the scale run script.  I have renamed the `--debug` flag to `--noop` in order to align with its functionality.  Let me know if there is anything else for me to address.

RandellP commented on docs/scale_run_script.md:
```diff
@@ -0,0 +1,71 @@
+  not be applied to this instance due to insufficient ram.
+* 5 cold start runs (provisioning on the first) with tune set to true on the
+  following:
+    * c5.xlarge: Starting with 2000 instances
```
I would avoid putting instance counts here so they don't get out of sync with the code... Maybe just add a line that the instance counts are defined at the top of the file or something.

johnduarte replied:
Doc updated to avoid using instance counts.

RandellP commented on docs/scale_run_script.md:
````diff
@@ -0,0 +1,71 @@
+be used with this flag.
+
+```
+./util/scale_run.sh -n -i foo 2019.1.0
````
you mention above that a jira ticket might be used... but then the example doesn't...

johnduarte replied:
doc updated to use JIRA ticket number

RandellP commented on docs/scale_run_script.md:
````diff
@@ -0,0 +1,71 @@
+
+This behaviour is executed via the following command:
+```
+./util/scale_run.sh 2019.1.0
````
If I understand it right, the "normal" way to run this is with the run-it set to the ticket number.  If so, having the first example a reader runs into being the one they should normally use would be helpful.

johnduarte replied:
I don't know if using the using the `--run-id` flag would be considered normal or not.   Regardless, my intention with the first example is to show the simplest way to invoke it.  The flags are described below and can then be appended by the user as needed.

RandellP replied:
Well, like say we hire someone new.  And they get the job of running the scale tests.  What command line would you expect them to end up running... That should be the first command line example seen in the doc (IMO).  Even if it means adding a new small section saying so.  I am only assuming that would include the ticket number as the run-id.  If not, and this is what you would expect them to run, then all good.

johnduarte replied:
This is a great discussion.

The workflow I used was to create a dedicated test runner for running the slv-414 runs. So, I didn't need a run identifier to scope them for that effort.  For the re-run work that I did, a simple time stamp was enough to 1) Ensure that I didn't overwrite run data; 2) Allowed be to keep track of the sequence of runs.  In the current version of the script, a timestamp is always included to meet these criteria.

I am curious to hear @billclaytor thoughts on the matter.

billclaytor commented on util/scale_run.sh:
```diff
@@ -0,0 +1,335 @@
+    echo "  -i|--run-id           Set a value to prepend to scale run directories"
+    echo "                        (default: '')"
+    echo "  --[no]tune            apply pe-tune to deployment"
+    echo "                        (default: 'true')"
```
Should this be "(default: 'tune')"?

johnduarte replied:
Yes.
Usage string updated.

billclaytor commented on util/scale_run.sh:
```diff
@@ -0,0 +1,335 @@
+# export PUPPET_GATLING_SCALE_TUNE_FORCE=true
+
+# AWS instance sizes to be tested on
+declare -a TRIAL_EC2_TYPES=("c5.large")
```
Aren't we using m5.large for the trial instance type?

johnduarte replied:
:man_facepalming: yes. Updated.

billclaytor commented on docs/scale_run_script.md:
```diff
@@ -0,0 +1,71 @@
+
+Out of the box, it will run the following:
+
+* 1 run of 100 instances on a c5.large instance for the trial license
```
Aren't we using m5.large for the trial instance type?

johnduarte replied:
:man_facepalming: Yes. Updated

billclaytor commented on util/scale_run.sh:
```diff
@@ -0,0 +1,335 @@
+fi
+PE_VERSION=${ARGS[0]}
+
+# Ensure PE_VERSION is semver
```
This doesn't seem to work when running with the -n option:
```
bill.claytor:~/RubymineProjects/_pr/SLV-414/gatling-puppet-load-test> ./util/scale_run.sh -n -i SLV-414 testing123
PE_MAJOR=testing123
PE_MINOR=testing123
PE_PATCH=testing123
PE_BUILD=testing123
```

johnduarte replied:
I have patched this bug.

Here are some tests
```
$ ./util/scale_run.sh -n 2019.1.1-rc0-207-g4e47830
PE_MAJOR=2019
PE_MINOR=1
PE_PATCH=1
PE_BUILD=-rc0-207-g4e47830
...
$ ./util/scale_run.sh -n 1.1.1
PE_MAJOR=1
PE_MINOR=1
PE_PATCH=1
PE_BUILD=
...
$ ./util/scale_run.sh -n 1.1
PE_MAJOR=
PE_MINOR=
PE_PATCH=
PE_BUILD=
Supplied PE_VERSION (1.1) is malformed
...
$ ./util/scale_run.sh -n 1.1.
PE_MAJOR=1
PE_MINOR=1
PE_PATCH=
PE_BUILD=
Supplied PE_VERSION (1.1.) is malformed
...
 $ ./util/scale_run.sh -n testing123
PE_MAJOR=
PE_MINOR=
PE_PATCH=
PE_BUILD=
Supplied PE_VERSION (testing123) is malformed
```
